lib/gc/Transforms/MergeAllocTickBased.cpp

```diff
@@ -0,0 +1,423 @@
+    for (auto &tr : trace) {
+      retTrace->traces.emplace_back(tr.trace);
+    }
+    ret.scopeTraces.emplace_back(std::move(retTrace));
+  }
+  // stablize the order of scopes for testing
+  std::stable_sort(
+      ret.scopeTraces.begin(), ret.scopeTraces.end(),
+      [](const std::unique_ptr<LifetimeTrace> &a,
+         const std::unique_ptr<LifetimeTrace> &b) {
+        int64_t aFirstSize = -1, bFirstSize = -1;
+        if (auto &traces = static_cast<TickTraceResult *>(a.get())->traces;
+            !traces.empty()) {
+          aFirstSize = traces.front().size;
+        }
+        if (auto &traces = static_cast<TickTraceResult *>(b.get())->traces;
+            !traces.empty()) {
+          bFirstSize = traces.front().size;
+        }
+        return aFirstSize < bFirstSize;
+      });
+  return ret;
+}
+
+FailureOr<MemoryTraceScopes>
+TickCollecter::operator()(Operation *root,
+                          const mlir::BufferViewFlowAnalysis &aliasAnaly,
+                          const MergeAllocationOptions &option) const {
+  TickCollecterStates s{aliasAnaly, option};
+  TickCollecter collecter;
+  auto result = root->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (failed(collecter.popScopeIfNecessary(&s, op))) {
+      return WalkResult::interrupt();
+    }
+    collecter.forwardTick(&s);
+    if (auto viewop = dyn_cast<ViewLikeOpInterface>(op)) {
+      collecter.onMemrefViews(&s, viewop);
+    } else if (op->hasTrait<OpTrait::ReturnLike>()) {
+      collecter.onReturnOp(&s, op);
+    } else if (isa<memref::AllocOp>(op)) {
+      collecter.onAllocOp(&s, op);
+    } else if (isa<memref::ExtractAlignedPointerAsIndexOp,
+                   memref::ExtractStridedMetadataOp>(op)) {
+      // TODO: should detect an op-trait which extracts pointer from memref
+      collecter.onExtractPointerOp(&s, op);
+    } else if (!isMemoryEffectFree(op)) {
+      // if the op has no memory effects, it don't contribute to liveness
+      collecter.onGeneralOp(&s, op);
+    }
+    if (op->getNumRegions() > 0 && !isa<func::FuncOp>(op)) {
+      // finally, if op is complex scope, push one ComplexScope
+      collecter.pushComplexScope(&s, op);
+    }
+    return WalkResult::advance();
+  });
+  if (result.wasInterrupted()) {
+    return failure();
+  }
+  if (failed(collecter.popScopeIfNecessary(&s, nullptr))) {
+    return failure();
+  }
+  if (option.checkOnly) {
+    for (auto &[alloc, tick] : s.allocTicks) {
+      auto allocscope = getAllocScope(&s, alloc);
+      alloc->setAttr(
+          "__mergealloc_lifetime",
+          DenseI64ArrayAttr::get(root->getContext(),
+                                 {reinterpret_cast<int64_t>(allocscope),
+                                  tick.firstAccess, tick.lastAccess}));
+      allocscope->setAttr(
+          "__mergealloc_scope",
+          IntegerAttr::get(mlir::IntegerType::get(root->getContext(), 64),
+                           reinterpret_cast<int64_t>(allocscope)));
```

Comment:
Better to use `constexpr StringLiteral mergeAllocXXXAttrName = "__mergealloc_xxx__"` instead of raw string, similar to upstream `__inplace_operands_attr__`, which would be friendly if we want to get attribute in the future.

Reply:
It is for testing only. The data is not actually passed via attr.